Add a couple of fuzzers for network_load_one and netdev_load_one

evverx opened the pull request:
These are the fuzzers that keep finding issues like https://github.com/systemd/systemd/issues/10615, https://github.com/systemd/systemd/issues/10628, https://github.com/systemd/systemd/issues/10629 and https://github.com/systemd/systemd/issues/10639.

evverx commented:
Just two tests failed on semaphoreci because `libsasan` there is too old. I'm not sure why @xnox didn't reply in https://github.com/systemd/systemd/pull/10190#issuecomment-425307072, so @martinpitt @mbiebl could you add `-Dslow-tests=true` to https://salsa.debian.org/systemd-team/systemd/blob/master/debian/rules?

yuwata commented:
LGTM. Let's merge this after #10618, #10633 and #10641 are merged.

yuwata commented:
> the following commands can be used to generate

@evverx I think parsing gperf file is better for that purpose. If I remember correctly, the man pages contain not all supported options. Also, for backward compatibility reason, `CheckSum` and `Checksum` are supporetd. But only `Checksum` is documented.

evverx commented:
@yuwata thank you! That's a great idea and I don't even know why I didn't think of it. It will be not only more accurate, but also a little bit easier in terms of regular expressions.

evverx commented:
It'd also be great if it was possible to automatically get a set of typical values each directive can accept. I could create wonderful dictionaries that would point the fuzzers in the right direction :-) It's not to say that now I point them in the wrong direction, though.

evverx commented:
Rebased on top of master to see how it goes. It's expected that the wireguard code still crashes the netdev fuzzer.

evverx commented:
I wrote a script that uses the *.gperf files:
```sh
perl -aF'/[\s,]+/' -ne '
if (my ($s, $d) = ($F[0] =~ /^([^\s\.]+)\.([^\s\.]+)$/)) { $d{$s}{$d} = 1; }
END { while (my ($key, $value) = each %d) {
    printf "[%s]\n%s\n", $key, join("\n", keys(%$value))
}}' src/network/networkd-network-gperf.gperf
```
Not to worry. I'm not going to try to push the script to the repository :-)

yuwata commented:
@evverx Could you rebase this PR?

evverx commented:
@yuwata yes, I'm on it. I should also probably change the data I feed to the fuzzers a bit now that https://github.com/systemd/systemd/pull/10653 was merged.

evverx commented:
I've just rebased the branch to see whether `semaphoreci` will become green. I haven't changed the data passed to the fuzzers yet.

evverx commented:
I replaced `AdActorSysPrio` with `AdActorSystemPriority` (though, frankly, I like AdActorSysPrio more), added two files containing all the sections and directives to help the fuzzers discover new paths faster and also modified `oss-fuzz.sh` a bit to make it possible to copy dictionaries along with options files.

evverx commented:
I was going to write that [the fuzzers compiled successfully](https://oss-fuzz-build-logs.storage.googleapis.com/index.html) when I received two letters from ClusterFuzz regarding bugs in the wireguard code. I don't think the issues are severe enough to hide them. @yuwata I can open a PR with both reproducers attached so that @Mic92 could take a look at and probably fix them. Plus, it would help to see whether https://github.com/systemd/systemd/pull/10668 really does what it advertises. What do you think?

yuwata commented:
Oh, I've just opened #10670. PTAL.

evverx commented:
Oh, as usual, you fixed both issues faster than I download the reproducers :-)

evverx commented:
@Mic92 I'd like to apologize for my last comment. In the end, those weren't bugs in the wrieguard code. It just so happened that ClusterFuzz apparently favoured the wireguard directives. The backtraces I looked at would have been different if some other directives had been chosen.

@yuwata thank you again for fixing this!
